README.md

```diff
@@ -184,16 +184,21 @@ MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install
 
 On Windows
 ```cmd
-set "VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Auxiliary\Build"
+REM The folowing line is for Visual Studio 2017 Enterprise only
```

Comment:
You missed `BuildTools`. It is also a product of VS 2017. But I don't think listing products here is a good idea. Actually, you can use `vswhere`. The usage is listed here: https://github.com/peterjc123/pytorch-scripts/blob/master/internal/check_deps.bat#L17-35. It is also acceptable if you just list them in the comments.

Reply:
> You missed `BuildTools`

By `BuildTools` do you mean
`VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\Tools\`
which is set after vcvarsall.bat invocation?

As for `vswhere` for Visual Studio 2017 only, well it is not really useful for the end user as it outputs a big list of info and that's it. But you will still need to mention in Readme `..\VC\Auxiliary\Build` as it is not presented in its output. Nonetheless, it might be used for obtaining `installationPath`, to which `\VC\Auxiliary\Build` should be appended (manual copy-paste, IMHO).

Reply:
> By `BuildTools` do you mean
`VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\Tools\`
which is set after vcvarsall.bat invocation?

No, I mean `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build`. It can be installed using the standalone installer. Please read more details [here](https://devblogs.microsoft.com/cppblog/introducing-the-visual-studio-build-tools/).

> As for `vswhere` for Visual Studio 2017 only, well it is not really useful for the end user as it outputs a big list of info and that's it. But you will still need to mention in Readme `..\VC\Auxiliary\Build` as it is not presented in its output. Nonetheless, it might be used for obtaining `installationPath`, to which `\VC\Auxiliary\Build` should be appended (manual copy-paste, IMHO).

Please read the script carefully (in the link). It also does what you said. VS 2017 isn't a problem because our code relies on it.

Reply:
1. I saw it in your script, but it looks like it is not presented in pytorch? Correct me if I wrong.
2. The name of the 'product' (Enterprise) has been already in Readme. I've just added the rest 2 (Community and Professional), and that's it. Otherwise, it looked like Enterprise is the only suitable 'product'.
So, what you want me to correct here? I didn't test with [Build tools](https://devblogs.microsoft.com/cppblog/introducing-the-visual-studio-build-tools/), so I couldn't write anything about it here.

Reply:
1. Yes, it is. It's not in because it is somehow too complex.
2. Instead of listing the paths here, just list the products here. And please add in `BuildTools`, whose path structure is just the same as the other products.

Reply:
1. Maybe a single command will suit us?

``FOR /f "usebackq tokens=*" %i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -property installationPath`) do SET VS150COMNTOOLS=%i\VC\Auxiliary\Build``

Additionally, we may ask a user to check: `echo %VS150COMNTOOLS%` or may not, cause the further call will check it: `call "%VS150COMNTOOLS%\vcvarsall.bat" x64`

Reply:
2. Will experiment with BuildTools and VC++ 2017 v141 toolset, but would prefer to mention it in Readme in another PR.

Reply:
Sounds good.

Reply:
1. Let's cram it in a single command:

``FOR /f "usebackq tokens=*" %i IN (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -property installationPath`) DO CALL "%i\VC\Auxiliary\Build\vcvarsall.bat" x64``

I've tested it from Command Prompt on Windows 7, 8, 8.1 and 10. Where both Visual Studio 2015 and 2017 were installed (Community and Professional).

Reply:
looks okay to me

Reply:
Wait, you 'd better add the argument `-products *` into this, which will also find buildtools.

Reply:
Are you sure about build tools? It looks like it is about products only.
```
  -products arg  One or more product IDs to find. Defaults to Community, Professional, and Enterprise.
                 Specify "*" by itself to search all product instances installed.
                 See https://aka.ms/vs/workloads for a list of product IDs.
```
Btw, the above `for` command should be testes on multiple Visual Studio 2017 installed. I'm going to install Community to check such config.

Reply:
I've tested this argument in real CI. Without it, it will only find `Community, Professional, and Enterprise` products.

Reply:
Could you please provide your output, because I've just successfully installed "VC++ 2017 version 15.4 v14.11 toolset", and adding `-products *` doesn't change anything in my case:

```
C:\Windows\system32>"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -products * -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional
```

Reply:
```cmd

C:\Users\Administrator\Desktop>"%ProgramFiles(x86)%\Microsoft Visual Studio\Inst
aller\vswhere.exe" -version [15^,16^) -products * -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community

C:\Users\Administrator\Desktop>"%ProgramFiles(x86)%\Microsoft Visual Studio\Inst
aller\vswhere.exe" -version [15^,16^) -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community

C:\Users\Administrator\Desktop>
```

Reply:
Hmmm... I See only "Professional", and there is no corresponding folder for BuildTools near it as well, even though I successfully installed "VC++ 2017 version 15.4 v14.11 toolset". Have I missed something concerning BuildTools?

Reply:
![02](https://user-images.githubusercontent.com/10241941/53571648-79fa3380-3b7a-11e9-8098-1e7b2ae9a595.png)

Reply:
No no no, by buildtools, I mean installing VS through this installer `https://download.visualstudio.microsoft.com/download/pr/f8db94a7-42dc-4413-b987-b21dcb0b1661/5e24cf2c187f46bde25c5515250fa9a0/vs_buildtools.exe`. It is also a product, just like `Professional` and it can be installed even if you have some other products like `Professional`.

Reply:
It looks like it is not shipped by Microsoft currently... Well, nonetheless, do we need BuildTools to be installed and mentioned in Readme? Or that is yet another Visual Studio Product like Community/Professional/Enterprise which might be used for building pytorch?

Reply:
> It looks like it is not shipped by Microsoft currently...

It's actually shipped by Microsoft in https://visualstudio.microsoft.com/downloads/.
![image](https://user-images.githubusercontent.com/9998726/53573038-99f41c00-3ba7-11e9-8a66-8e165ad546df.png)

> Or that is yet another Visual Studio Product like Community/Professional/Enterprise which might be used for building pytorch?

When using `-products *`, we are using all the possible products. So we don't have to care about it.

Reply:
Ok, thanks, I missed BuildTools on Microsoft.com.
Now we should choose, I think, one product in order not to get multiple environment initialisations in the loop.

```
C:\Windows\system32>for /f "usebackq tokens=*" %i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [14^,16^) -products * -property installationPath`) do call "%i\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11

C:\Windows\system32>call "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Windows\System32>call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
```

Reply:
Agreed.

Reply:
I think only the user can choose the product in case of multiple products installed. Searching an appropriate product with the needed toolset is the work for the script, not for a couple of commands. We can check such situation with multiple products and let him/her choose, otherwise (if only one VS 2017 product is installed) do the job seamlessly without user engagement. What do you think?

Reply:
One more option is to use "-latest":

`-latest        Return only the newest version and last installed.`

In this case, a single product will always be chosen. If the product will not suit - the user will have to update it and/or install the needed toolset.

I like this option more.